### ISSUE TITLE
In-place update on the original value tensor should also update version counter of sparse tensor's values_ tensor

### ISSUE THREAD

**Issue by yf225** [pytorch/pytorch#22778]
Currently, in-place updating the original value tensor after it's been used to construct a sparse tensor doesn't update version counter of the sparse tensor's `values_` tensor. Consider the following example:
```python
ind = torch.tensor([[0],[1]]).add_(1).sub_(1)
values = torch.tensor([1.]).add_(1).add_(1).sub_(1).sub_(1)
c = torch.sparse_coo_tensor(ind, values).requires_grad_()
grad = torch.sparse_coo_tensor(ind.clone(), values.clone())
d = c.mul(c)
values.add_(1)
d.backward(grad)  # This should have thrown an error saying there is version mismatch error for c
```
The reason is that we don't save the values tensor for backward in the sparse tensor constructor: 
```cpp
Tensor VariableType::_sparse_coo_tensor_with_dims_and_tensors(int64_t sparse_dim, int64_t dense_dim, IntArrayRef size, const Tensor & indices, const Tensor & values, const TensorOptions & options) {
  RECORD_FUNCTION("_sparse_coo_tensor_with_dims_and_tensors", std::vector<c10::IValue>({indices, values}), Function::peek_at_next_sequence_nr());
  auto& indices_ = unpack(indices, "indices", 3);
  auto& values_ = unpack(values, "values", 4);
  auto options_ = TensorOptions(options).is_variable(false);
  check_no_requires_grad(indices, "indices");
  std::shared_ptr<SparseCooTensorWithDimsAndTensorsBackward> grad_fn;
  if (compute_requires_grad( values )) {
    grad_fn = std::shared_ptr<SparseCooTensorWithDimsAndTensorsBackward>(new SparseCooTensorWithDimsAndTensorsBackward(), deleteFunction);
    grad_fn->set_next_edges(collect_next_edges( values ));
    grad_fn->indices_ = SavedVariable(indices, false);
    grad_fn->values_sizes = values.sizes().vec();
    // We should also save `values` here
  }
  ...
}
```

**Comment by SsnL** [comment]
I’m pretty sure the sparse tensor share version counter with the value tensor.

**Comment by albanD** [comment]
Well, there is a typo [here](https://github.com/pytorch/pytorch/blob/c6fa6d82aebc1dcd7561ea29ec5d41c5a211bae1/tools/autograd/gen_autograd.py#L61) (missing the `_` at the beginning of the function)  so at the moment, they do not share version counter.

But marking the sparse Tensor can lead to tricky situations. In particular if inplace modifications are done, it will try to rewrite the graph as if it was just a view, and will ignore the fact that something is sparse. Leading to wrong gradients.
We need to either:
- forbid inplace operation on the resulting sparse Tensor and original values.
- only share the version counter but not as a view. And gradients will be wrong if inplace are done both on the sparse tensor and the input values.

Any opinion?
